File: src/common/hash.zig
/// hash.zig — Generic cryptographic hashing utilities
///
/// Provides allocator-friendly SHA-256 helpers and a content+model hash
/// suitable for embedding cache keys.  All functions are free of
/// guidance-domain types so they can be reused by any Zig tool.
const std = @import("std");

/// Converts input data into a SHA256 hexadecimal string using an allocator.
pub fn sha256Hex(allocator: std.mem.Allocator, data: []const u8) ![]const u8 {
    var hash_out: [32]u8 = undefined;
    std.crypto.hash.sha2.Sha256.hash(data, &hash_out, .{});
    const hex = std.fmt.bytesToHex(hash_out, .lower);
    const result = try allocator.alloc(u8, hex.len);
    @memcpy(result, &hex);
    return result;
}

/// Computes a 16-bit hash combining content and model data.
pub fn contentHashWithModel(content: []const u8, model: []const u8) [16]u8 {
    var hasher = std.crypto.hash.sha2.Sha256.init(.{});
    hasher.update(model);
    hasher.update("\x00");
    hasher.update(content);
    var digest: [32]u8 = undefined;
    hasher.final(&digest);

    var result: [16]u8 = undefined;
    const hex_chars = "0123456789abcdef";
    for (0..8) |i| {
        result[i * 2] = hex_chars[digest[i] >> 4];
        result[i * 2 + 1] = hex_chars[digest[i] & 0x0f];
    }
    return result;
}

// =============================================================================
// Multi-algorithm hashing — from coral/src/common/hash.zig
// =============================================================================

pub const HashAlgorithm = enum {
    sha256,
    sha512,
    blake3,

    pub fn digestLength(self: HashAlgorithm) usize {
        return switch (self) {
            .sha256 => 32,
            .sha512 => 64,
            .blake3 => 32,
        };
    }
};

/// Computes a hash for a file using the provided allocator and algorithm, returning the resulting hash value.
pub fn hashFile(
    allocator: std.mem.Allocator,
    path: []const u8,
    algorithm: HashAlgorithm,
) ![]const u8 {
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();

    var buf: [65536]u8 = undefined;

    return switch (algorithm) {
        .sha256 => hashFileGeneric(allocator, &buf, file, std.crypto.hash.sha2.Sha256.init(.{})),
        .sha512 => hashFileGeneric(allocator, &buf, file, std.crypto.hash.sha2.Sha512.init(.{})),
        .blake3 => hashFileGeneric(allocator, &buf, file, std.crypto.hash.Blake3.init(.{})),
    };
}

fn hashFileGeneric(
    allocator: std.mem.Allocator,
    buf: *[65536]u8,
    file: std.fs.File,
    hasher: anytype,
) ![]const u8 {
    const H = @TypeOf(hasher);
    var h = hasher;
    while (true) {
        const n = file.read(buf) catch break;
        if (n == 0) break;
        h.update(buf[0..n]);
    }
    var out: [H.digest_length]u8 = undefined;
    h.final(&out);
    const hex = std.fmt.bytesToHex(out, .lower);
    return allocator.dupe(u8, &hex);
}

pub const BatchHashResult = struct {
    path: []const u8,
    hash: ?[]const u8,
    error_msg: ?[]const u8,
};

/// Processes multiple paths with a hashing algorithm, returning results via callback.
pub fn hashBatch(
    allocator: std.mem.Allocator,
    paths: []const []const u8,
    algorithm: HashAlgorithm,
    progress_callback: ?*const fn (usize, usize) void,
) ![]BatchHashResult {
    var results = try allocator.alloc(BatchHashResult, paths.len);
    errdefer allocator.free(results);

    for (paths, 0..) |path, i| {
        if (progress_callback) |cb| cb(i, paths.len);

        results[i] = .{ .path = path, .hash = null, .error_msg = null };

        const hash = hashFile(allocator, path, algorithm) catch |err| {
            results[i].error_msg = try std.fmt.allocPrint(allocator, "{}", .{err});
            continue;
        };
        results[i].hash = hash;
    }

    return results;
}

/// FNV-1a 64-bit hash — non-cryptographic, stable, fast, suitable for cache keys.
/// Uses the standard FNV-1a offset basis and prime.
pub fn fnv1a64(input: []const u8) u64 {
    const FNV_OFFSET: u64 = 14695981039346656037;
    const FNV_PRIME: u64 = 1099511628211;
    var h: u64 = FNV_OFFSET;
    for (input) |byte| {
        h ^= byte;
        h *%= FNV_PRIME;
    }
    return h;
}

/// Converts a byte slice into a Zig hash using the provided algorithm.
pub fn hashString(
    allocator: std.mem.Allocator,
    data: []const u8,
    algorithm: HashAlgorithm,
) ![]const u8 {
    return switch (algorithm) {
        .sha256 => computeHashHex(allocator, data, std.crypto.hash.sha2.Sha256),
        .sha512 => computeHashHex(allocator, data, std.crypto.hash.sha2.Sha512),
        .blake3 => computeHashHex(allocator, data, std.crypto.hash.Blake3),
    };
}

fn computeHashHex(allocator: std.mem.Allocator, data: []const u8, HashType: type) ![]const u8 {
    var out: [HashType.digest_length]u8 = undefined;
    HashType.hash(data, &out, .{});
    const hex = std.fmt.bytesToHex(out, .lower);
    return allocator.dupe(u8, &hex);
}

/// Computes a 32-byte Blake3 hash from the provided data slice.
pub fn blake3Hash(data: []const u8) [32]u8 {
    var out: [32]u8 = undefined;
    std.crypto.hash.Blake3.hash(data, &out, .{});
    return out;
}

// =============================================================================
// In-memory query cache — FNV-keyed, session-scoped
// =============================================================================

/// In-memory cache for query results, keyed by FNV-1a64 hash of the query string.
/// Intended for session-scoped use (e.g. `guidance serve`) to avoid repeating
/// the expensive search + synthesis pipeline for identical queries.
///
/// Thread-safety: NOT thread-safe — query execution is single-threaded per VISION.md.
/// Eviction: when `entries.count() >= max_entries`, the oldest-iterated entry is dropped
/// (first-found, not true LRU — sufficient for the expected small working set).
pub const QueryCache = struct {
    allocator: std.mem.Allocator,
    entries: std.StringHashMapUnmanaged(CacheEntry),
    max_entries: usize,

    pub const CacheEntry = struct {
        /// Owned lowercase copy of the original query.
        lower_query: []u8,
        /// Owned copy of the synthesised result summary.
        result_summary: []u8,
        timestamp: i128,
    };

    pub fn init(allocator: std.mem.Allocator) QueryCache {
        return .{
            .allocator = allocator,
            .entries = .empty,
            .max_entries = 256,
        };
    }

    pub fn deinit(self: *QueryCache) void {
        var it = self.entries.iterator();
        while (it.next()) |kv| {
            self.allocator.free(kv.key_ptr.*);
            self.allocator.free(kv.value_ptr.lower_query);
            self.allocator.free(kv.value_ptr.result_summary);
        }
        self.entries.deinit(self.allocator);
    }

    /// Returns the cached summary for `query`, or `null` on a cache miss.
    pub fn get(self: *QueryCache, query: []const u8) ?[]const u8 {
        var key_buf: [17]u8 = undefined;
        const h = fnv1a64(query);
        const key = std.fmt.bufPrint(&key_buf, "{x:0>16}", .{h}) catch return null;
        const entry = self.entries.get(key) orelse return null;
        return entry.result_summary;
    }

    /// Stores `summary` under `query`. Evicts one entry if at capacity.
    /// Caller retains ownership of `summary`; an internal copy is made.
    pub fn put(self: *QueryCache, query: []const u8, summary: []const u8) !void {
        if (self.entries.count() >= self.max_entries) {
            var evict_it = self.entries.iterator();
            if (evict_it.next()) |kv| {
                const old_key = kv.key_ptr.*;
                const old_val = kv.value_ptr.*;
                _ = self.entries.remove(old_key);
                self.allocator.free(old_key);
                self.allocator.free(old_val.lower_query);
                self.allocator.free(old_val.result_summary);
            }
        }

        const h = fnv1a64(query);
        const key = try std.fmt.allocPrint(self.allocator, "{x:0>16}", .{h});
        errdefer self.allocator.free(key);

        const lower_query = try std.ascii.allocLowerString(self.allocator, query);
        errdefer self.allocator.free(lower_query);

        const result_summary = try self.allocator.dupe(u8, summary);
        errdefer self.allocator.free(result_summary);

        const gop = try self.entries.getOrPut(self.allocator, key);
        if (gop.found_existing) {
            // Update existing entry; the key slot in the map already owns its key.
            self.allocator.free(key);
            self.allocator.free(gop.value_ptr.lower_query);
            self.allocator.free(gop.value_ptr.result_summary);
        } else {
            gop.key_ptr.* = key;
        }
        gop.value_ptr.* = .{
            .lower_query = lower_query,
            .result_summary = result_summary,
            .timestamp = std.time.nanoTimestamp(),
        };
    }
};

/// Converts input data into a hexadecimal string using the Blake3 algorithm.
pub fn blake3Hex(allocator: std.mem.Allocator, data: []const u8) ![]const u8 {
    const h = blake3Hash(data);
    const hex = std.fmt.bytesToHex(h, .lower);
    return allocator.dupe(u8, &hex);
}

pub const HashState = struct {
    algorithm: HashAlgorithm,
    sha256: ?std.crypto.hash.sha2.Sha256 = null,
    sha512: ?std.crypto.hash.sha2.Sha512 = null,
    blake3_h: ?std.crypto.hash.Blake3 = null,

    pub fn init(algorithm: HashAlgorithm) HashState {
        return switch (algorithm) {
            .sha256 => .{ .algorithm = algorithm, .sha256 = std.crypto.hash.sha2.Sha256.init(.{}) },
            .sha512 => .{ .algorithm = algorithm, .sha512 = std.crypto.hash.sha2.Sha512.init(.{}) },
            .blake3 => .{ .algorithm = algorithm, .blake3_h = std.crypto.hash.Blake3.init(.{}) },
        };
    }

    pub fn update(self: *HashState, data: []const u8) void {
        switch (self.algorithm) {
            .sha256 => self.sha256.?.update(data),
            .sha512 => self.sha512.?.update(data),
            .blake3 => self.blake3_h.?.update(data),
        }
    }

    /// Finalise the hash and return an allocator-owned hex string; caller must free.
    pub fn final(self: *HashState, allocator: std.mem.Allocator) ![]const u8 {
        return switch (self.algorithm) {
            .sha256 => finalizeHashHex(allocator, &self.sha256.?),
            .sha512 => finalizeHashHex(allocator, &self.sha512.?),
            .blake3 => finalizeHashHex(allocator, &self.blake3_h.?),
        };
    }
};

fn finalizeHashHex(allocator: std.mem.Allocator, hasher: anytype) ![]const u8 {
    const H = @TypeOf(hasher.*);
    var out: [H.digest_length]u8 = undefined;
    hasher.final(&out);
    const hex = std.fmt.bytesToHex(out, .lower);
    return allocator.dupe(u8, &hex);
}

// =============================================================================
// Tests
// =============================================================================

File: src/common/src/hash.rs
use blake3::Hasher;
use sha2::{Digest, Sha256, Sha512};
use std::io::Read;
use std::path::Path;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HashAlgorithm {
    Sha256,
    Sha512,
    Blake3,
}

impl HashAlgorithm {
    pub fn digest_length(self) -> usize {
        match self {
            HashAlgorithm::Sha512 => 64,
            HashAlgorithm::Sha256 | HashAlgorithm::Blake3 => 32,
        }
    }
}

pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let result = hasher.finalize();
    hex_encode(&result)
}

pub fn blake3_hash(data: &[u8]) -> [u8; 32] {
    blake3::hash(data).into()
}

pub fn blake3_hex(data: &[u8]) -> String {
    blake3::hash(data).to_hex().to_string()
}

pub fn fnv1a64(input: &[u8]) -> u64 {
    const OFFSET: u64 = 14_695_981_039_346_656_037;
    const PRIME: u64 = 1_099_511_628_211;
    let mut hash: u64 = OFFSET;
    for &byte in input {
        hash ^= u64::from(byte);
        hash = hash.wrapping_mul(PRIME);
    }
    hash
}

pub fn content_hash_with_model(content: &str, model: &str) -> [u8; 16] {
    let mut hasher = Sha256::new();
    hasher.update(model.as_bytes());
    hasher.update(b"\x00");
    hasher.update(content.as_bytes());
    let result = hasher.finalize();
    let mut out = [0u8; 16];
    out.copy_from_slice(&result[..16]);
    out
}

pub struct HashState {
    #[allow(dead_code)]
    algorithm: HashAlgorithm,
    sha256: Option<Sha256>,
    sha512: Option<Sha512>,
    blake3: Option<Hasher>,
}

impl HashState {
    pub fn new(algorithm: HashAlgorithm) -> Self {
        match algorithm {
            HashAlgorithm::Sha256 => Self {
                algorithm,
                sha256: Some(Sha256::new()),
                sha512: None,
                blake3: None,
            },
            HashAlgorithm::Sha512 => Self {
                algorithm,
                sha256: None,
                sha512: Some(Sha512::new()),
                blake3: None,
            },
            HashAlgorithm::Blake3 => Self {
                algorithm,
                sha256: None,
                sha512: None,
                blake3: Some(Hasher::new()),
            },
        }
    }

    pub fn update(&mut self, data: &[u8]) {
        if let Some(ref mut h) = self.sha256 {
            h.update(data);
        }
        if let Some(ref mut h) = self.sha512 {
            h.update(data);
        }
        if let Some(ref mut h) = self.blake3 {
            h.update(data);
        }
    }

    pub fn finalize(&self) -> Vec<u8> {
        if let Some(ref h) = self.sha256 {
            h.clone().finalize().to_vec()
        } else if let Some(ref h) = self.sha512 {
            h.clone().finalize().to_vec()
        } else if let Some(ref h) = self.blake3 {
            h.clone().finalize().as_bytes().to_vec()
        } else {
            Vec::new()
        }
    }

    pub fn digest_hex(&self) -> String {
        hex_encode(&self.finalize())
    }
}

#[derive(Debug, Clone)]
pub struct BatchHashResult {
    pub path: String,
    pub hash: Option<String>,
}

pub fn hash_file(path: &Path, algorithm: HashAlgorithm) -> std::io::Result<String> {
    let mut file = std::fs::File::open(path)?;
    let mut hasher = HashState::new(algorithm);
    let mut buf = [0u8; 65536];
    loop {
        let n = file.read(&mut buf)?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    Ok(hasher.digest_hex())
}

pub fn hash_batch(paths: &[std::path::PathBuf], algorithm: HashAlgorithm) -> Vec<BatchHashResult> {
    paths
        .iter()
        .map(|path| {
            let hash = hash_file(path, algorithm).ok();
            BatchHashResult {
                path: path.to_string_lossy().to_string(),
                hash,
            }
        })
        .collect()
}

fn hex_encode(data: &[u8]) -> String {
    data.iter()
        .map(|b| format!("{b:02x}"))
        .collect::<Vec<_>>()
        .concat()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sha256_hex_correct_length() {
        let h = sha256_hex(b"hello");
        assert_eq!(h.len(), 64);
    }

    #[test]
    fn sha256_hex_deterministic() {
        assert_eq!(sha256_hex(b"hello"), sha256_hex(b"hello"));
    }

    #[test]
    fn sha256_hex_differs_for_different_inputs() {
        assert_ne!(sha256_hex(b"hello"), sha256_hex(b"world"));
    }

    #[test]
    fn blake3_hash_length() {
        let h = blake3_hash(b"test");
        assert_eq!(h.len(), 32);
    }

    #[test]
    fn blake3_hex_length() {
        let h = blake3_hex(b"test");
        assert_eq!(h.len(), 64);
    }

    #[test]
    fn fnv1a64_basic() {
        let h = fnv1a64(b"hello");
        assert_ne!(h, 0);
        assert_eq!(fnv1a64(b"hello"), fnv1a64(b"hello"));
    }

    #[test]
    fn digest_length_values() {
        assert_eq!(HashAlgorithm::Sha256.digest_length(), 32);
        assert_eq!(HashAlgorithm::Sha512.digest_length(), 64);
        assert_eq!(HashAlgorithm::Blake3.digest_length(), 32);
    }

    #[test]
    fn hash_state_incremental_sha256() {
        let mut state = HashState::new(HashAlgorithm::Sha256);
        state.update(b"hello ");
        state.update(b"world");
        let hex = state.digest_hex();
        assert_eq!(hex.len(), 64);
    }

    #[test]
    fn content_hash_deterministic() {
        let h1 = content_hash_with_model("hello", "model");
        let h2 = content_hash_with_model("hello", "model");
        assert_eq!(h1, h2);
    }

    #[test]
    fn content_hash_model_sensitive() {
        let h1 = content_hash_with_model("hello", "model-a");
        let h2 = content_hash_with_model("hello", "model-b");
        assert_ne!(h1, h2);
    }

    #[test]
    fn hash_file_small() {
        let dir = tempfile::TempDir::new().unwrap();
        let path = dir.path().join("test.txt");
        std::fs::write(&path, b"hello").unwrap();
        let hash = hash_file(&path, HashAlgorithm::Sha256).unwrap();
        assert_eq!(hash.len(), 64);
    }

    #[test]
    fn hash_file_nonexistent() {
        let result = hash_file(Path::new("/nonexistent/file.txt"), HashAlgorithm::Sha256);
        assert!(result.is_err());
    }

    #[test]
    fn hash_batch_all_succeed() {
        let dir = tempfile::TempDir::new().unwrap();
        let p1 = dir.path().join("a.txt");
        let p2 = dir.path().join("b.txt");
        std::fs::write(&p1, b"data1").unwrap();
        std::fs::write(&p2, b"data2").unwrap();
        let results = hash_batch(&[p1, p2], HashAlgorithm::Sha256);
        assert_eq!(results.len(), 2);
        assert!(results[0].hash.is_some());
        assert!(results[1].hash.is_some());
    }

    #[test]
    fn hash_batch_some_fail() {
        let dir = tempfile::TempDir::new().unwrap();
        let p1 = dir.path().join("exists.txt");
        let p2 = dir.path().join("missing.txt");
        std::fs::write(&p1, b"data").unwrap();
        let results = hash_batch(&[p1, p2], HashAlgorithm::Sha256);
        assert_eq!(results.len(), 2);
        assert!(results[0].hash.is_some());
        assert!(results[1].hash.is_none());
    }

    #[test]
    fn hash_state_incremental_sha512() {
        let mut state = HashState::new(HashAlgorithm::Sha512);
        state.update(b"test data");
        let hex = state.digest_hex();
        assert_eq!(hex.len(), 128);
    }

    #[test]
    fn hash_state_incremental_blake3() {
        let mut state = HashState::new(HashAlgorithm::Blake3);
        state.update(b"test data");
        let hex = state.digest_hex();
        assert_eq!(hex.len(), 64);
    }
}
